pyuvdata/utils.py

```diff
@@ -1518,27 +1529,260 @@ def uvcalibrate(
+    override_time_check : bool
+        Option to override the check that times match between the UVCal and UVData
+        objects. Only used in UVCal has only one time.
+    override_ant_check : bool
+        Option to override the check that all antennas with data on the UVData
+        object are present in the UVCal object. If this option is set to True,
+        uvcalibrate will proceed without erroring and data for antennas without
+        calibrations will be flagged.
 
     Returns
     -------
     UVData, optional
         Returns if not inplace
 
     """
-    # deepcopy for not inplace
+    # copy for not inplace
     if not inplace:
-        uvdata = copy.deepcopy(uvdata)
+        uvdata = uvdata.copy()
+
+    # Check whether the UVData antennas *that have data associated with them*
+    # have associated data in the UVCal object
+    uvdata_unique_nums = np.unique(np.append(uvdata.ant_1_array, uvdata.ant_2_array))
+    uvdata.antenna_names = np.asarray(uvdata.antenna_names)
+    uvdata_used_antnames = np.array(
+        [
+            uvdata.antenna_names[np.where(uvdata.antenna_numbers == antnum)][0]
+            for antnum in uvdata_unique_nums
+        ]
+    )
+    uvcal_unique_nums = np.unique(uvcal.ant_array)
+    uvcal.antenna_names = np.asarray(uvcal.antenna_names)
+    uvcal_used_antnames = np.array(
+        [
+            uvcal.antenna_names[np.where(uvcal.antenna_numbers == antnum)][0]
+            for antnum in uvcal_unique_nums
+        ]
+    )
+
+    ant_arr_match = uvcal_used_antnames.tolist() == uvdata_used_antnames.tolist()
+
+    if not ant_arr_match:
+        # check more carefully
+        for this_ant_name in uvdata_used_antnames:
+            wh_ant_match = np.nonzero(uvcal_used_antnames == this_ant_name)
+            if wh_ant_match[0].size == 0:
+                # old behavior only required that antenna numbers were present,#
+                # not names. Check numbers, if they are present issue
+                # deprecation warning
+                uvdata_ant_num = uvdata.antenna_numbers[
+                    np.where(uvdata.antenna_names == this_ant_name)[0][0]
+                ]
+                if override_ant_check:
+                    warnings.warn(
+                        f"Antenna {this_ant_name} has data on UVData but not on UVCal. "
+                        "override_ant_check is True, so the data associated with "
+                        "this antenna will be flagged."
+                    )
+                elif uvdata_ant_num in uvcal_unique_nums:
+                    warnings.warn(
+                        f"Antenna {this_ant_name} has data on UVData but not on UVCal. "
+                        f"Its antenna number is {uvdata_ant_num}, which is present "
+                        "on UVCal. Currently the data will be calibrated using the "
+                        "matching antenna number, but that will be deprecated in "
```

Comment:
hmm, I take your point. The problem is that there are a lot of existing files that this will be broken for (because until very recently hera_cal was overwriting the antenna names). What if I change this to check all antennas and only use the number if none of the antenna names match?

Reply:
To be sure I understand the proposed change:
- If no names match, use numbers
- If some names match, flag those that don't and use names for those that do.
- If all names match, use names

This sounds better to me. At least what is returned is self-consistent.

Reply:
ok, I made this change.